joomla/joomla-cms
# [4.1] Log deprecated only when explicitly requested

### Summary of Changes
Try to fix performance issue #30997.
~I have added proxy method `Log::deprecated()`, to use for deprecation log, which doing log only when `log_deprecated` is On.~

~However here a little caveats:
First: The patch add Application dependency to Log class. 
Second: The method may be called when Application is not initialized, then the method will do nothing, such logs may be missed in log output. This affect logs from bootup process. (Actually this is true for all bootup logs)~

Use of `trigger_error()` for log for deprecated messages.

@wilsonge @HLeithner please review

### Testing Instructions
Apply patch.
Enable debug.
In global config _enable_ "Log Deprecated API"
In the debug plugin _disable_ "Log Entries"

Open any page (example /administrator/index.php?option=com_config) 
check Debug Bar and note rendering time,

Then in global config _disable_ "Log Deprecated API", refresh the same page,
check Debug Bar and note rendering time,

### Actual result BEFORE applying this Pull Request
Debug Bar does not show logs but rendering time still huge.
At administrator/index.php?option=com_config I got around **3s**


### Expected result AFTER applying this Pull Request
Debug Bar does not show logs and rendering time is smaller.
At administrator/index.php?option=com_config I got around **270ms**


### Documentation Changes Required
none
